### PR TITLE
Warmup phase for faithfulness loss

### PR DESCRIPTION
**Pull request by leesharkey**
## Description
We don't have a great scheme for initializing the SPD parameters such that they start at a reasonable scale compared with the target model parameters. After all, the target model parameters might have quite different scales at different parts of the model, which means that the initial parameters might be off by different amounts in different parts of the model. By having a warmup for the SPD parameters, we remove one source of potential issues with our runs, since it's more likely that the parameters start in a regime that isn't bad for learning. 

## Related Issue
None

## How Has This Been Tested?
I've run it with TMS 5 -2 and ss_llama_single_gpu. Also it passes all tests, some of which use it. 

## Does this PR introduce a breaking change?
No.

### CODE REVIEW

**Comment by leesharkey** [comment]
I'm playing around with the code and seeing the effects of the faithfulness warmup on quantities like subcomponent norm and their cosine similarities with the target model. 

The main observation is that larger weight decay increases the tendency for the subcomponents to align with the target model (as expected). Another observation is that for tms5-2-id, a seed where the model fails to learn anything useful (seed=1) actually succeeds when we use the faithfulness warmup. I haven't thoroughly investigated whether this is due to the change in the initial weights, but i strongly suspect it is, and it points toward (but is definitely not conclusive evidence for) initialization being the cause of seed sensitivity. 

I don't include this code (from `run_spd.py`) in the branch, but I'm posting it here for posterity in case anyone wants to return to measuring these quantities during the faithfulness warmup training loop:

```
        for warmup_step in range(config.warmup_steps):
            warmup_optimizer.zero_grad()
            weight_deltas = component_model.calc_weight_deltas()
            faithfulness_loss = calc_faithfulness_loss(weight_deltas, device)
            faithfulness_loss.backward()
            warmup_optimizer.step()
            
            # Log faithfulness loss and component L2 norms
            # for comp_name, component in component_model.components.items():
            #     # Calculate frobenius norm of each rank one component made of U and V
            #     norm = torch.norm(component.U, dim=1) * torch.norm(component.V, dim=0)
            #     # Log the mean, min, max, and std of the norms
            #     mean_norm = norm.mean().item()
            #     min_norm = norm.min().item()
            #     max_norm = norm.max().item()
            #     std_norm = norm.std().item()

            #     # Calculate the cosine similarity between the component and the target weight
            #     target_weight = component_model.target_weight(comp_name)
            #     target_weight_flat = target_weight.reshape(-1)
            #     rank_one_matrices = einops.einsum(component_model.components[comp_name].V, component_model.components[comp_name].U, "d_in C, C d_out -> C d_in d_out")
            #     component_weight = rank_one_matrices.reshape(component_model.components[comp_name].C, -1)
            #     cosine_similarity = torch.nn.functional.cosine_similarity(component_weight, target_weight_flat, dim=1).mean().item()
            #     logger.info(f"{comp_name}: mean:{mean_norm:.6f}, min:{min_norm:.6f}, max:{max_norm:.6f}, std:{std_norm:.6f}, cosine similarity:{cosine_similarity:.6f}")

```

**Review by danbraunai — CHANGES_REQUESTED** [review]
Logic looks good.

I don't like the idea of adding a bunch of config arguments when implementing a feature that requires investigation. I definitely don't think we should be running with the faithfulness warmup enabled in all the configs when we haven't verified that it works just as well or better on all toy models and ideally it has some promise on LMs (this would be shown with links to evals in the description). Before then, I think we just keep this is a branch and not put it in main.

Thoughts?

File: spd/run_spd.py [hunk 1]
```diff
@@ -45,6 +45,48 @@
...
+    Args:
+        component_model: The component model to warm up
+        component_params: List of component parameters to optimize
+        config: Configuration object containing warmup settings
+        device: Device to run on
+    """
+
+    logger.info(
+        f"Starting faithfulness warmup phase with {config.faithfulness_warmup_steps} steps at lr={config.faithfulness_warmup_lr}"
+    )
```
[the remark this answers is not in the document]
Comment (leesharkey) [reply spd/run_spd.py:65]: It now looks like
```
Starting faithfulness warmup phase with 200 steps at lr=0.01
Faithfulness warmup step 1/200; Faithfulness loss: 0.007499937
Faithfulness warmup step 101/200; Faithfulness loss: 0.000000064
Faithfulness warmup step 200/200; Faithfulness loss: 0.000000001
Faithfulness warmup completed.
``

**Review by danbraunai — APPROVED** [review]
Sweet, looks good. Code looks clean and helpful to have the evaluations in PR comments. It's not immediately obvious what is being swept over in the multiple reports that you posted, but it at least looks like it doesn't hurt when I sampled a couple of them.

You can merge when you want.
